common/hashing.h

```diff
@@ -482,67 +546,110 @@ static_assert(std::has_unique_object_representations_v<int32_t>);
+                     objs.size() * sizeof(T)),
+      seed);
+}
 
 template <typename T>
-  requires NullPtrOrHasUniqueObjectRepresentations<T>
+  requires HashAsRawDataType<T>
 inline auto CarbonHashValue(const T& value, uint64_t seed) -> HashCode {
   Hasher hasher(seed);
-  hasher.Hash(MapNullPtrToVoidPtr(value));
+  hasher.Hash(MapToRawDataType(value));
   return static_cast<HashCode>(hasher);
 }
 
 template <typename... Ts>
-  requires(... && NullPtrOrHasUniqueObjectRepresentations<Ts>)
+  requires(... && HashAsRawDataType<Ts>)
 inline auto CarbonHashValue(const std::tuple<Ts...>& value, uint64_t seed)
     -> HashCode {
   Hasher hasher(seed);
   std::apply(
-      [&](const auto&... args) { hasher.Hash(MapNullPtrToVoidPtr(args)...); },
+      [&](const auto&... args) { hasher.Hash(MapToRawDataType(args)...); },
       value);
   return static_cast<HashCode>(hasher);
 }
 
 template <typename T, typename U>
-  requires NullPtrOrHasUniqueObjectRepresentations<T> &&
-           NullPtrOrHasUniqueObjectRepresentations<U> &&
-           (sizeof(T) <= sizeof(uint64_t) && sizeof(U) <= sizeof(uint64_t))
+  requires HashAsRawDataType<T> && HashAsRawDataType<U>
 inline auto CarbonHashValue(const std::pair<T, U>& value, uint64_t seed)
     -> HashCode {
-  return CarbonHashValue(std::tuple(value.first, value.second), seed);
+  Hasher hasher(seed);
+  hasher.Hash(MapToRawDataType(value.first), MapToRawDataType(value.second));
+  return static_cast<HashCode>(hasher);
 }
 
 template <typename T>
-  requires std::has_unique_object_representations_v<T>
+  requires HasCustomCarbonHashValue<T>
 inline auto CarbonHashValue(llvm::ArrayRef<T> objs, uint64_t seed) -> HashCode {
-  return CarbonHashValue(
-      llvm::ArrayRef(reinterpret_cast<const std::byte*>(objs.data()),
-                     objs.size() * sizeof(T)),
-      seed);
+  Hasher hasher(seed);
+  // Just use a simple loop. This isn't an especially optimized approach and we
+  // can enhance this if it ever proves important.
+  for (const T& obj : objs) {
+    hasher.Hash(HashValue(obj));
+  }
+  hasher.Hash(objs.size());
+  return static_cast<HashCode>(hasher);
+}
+
+template <typename... Ts>
+  requires(... && HasCustomCarbonHashValue<Ts>)
+inline auto CarbonHashValue(const std::tuple<Ts...>& value, uint64_t seed)
+    -> HashCode {
+  Hasher hasher(seed);
+  // No need to propagate the seed below as we've already incorporated it.
+  std::apply([&](const auto&... args) { hasher.Hash(HashValue(args)...); },
+             value);
+  return static_cast<HashCode>(hasher);
+}
+
+template <typename T, typename U>
+  requires HasCustomCarbonHashValue<T> && HasCustomCarbonHashValue<U>
+inline auto CarbonHashValue(const std::pair<T, U>& value, uint64_t seed)
+    -> HashCode {
+  Hasher hasher(seed);
+  // No need to propagate the seed below as we've already incorporated it.
+  hasher.Hash(HashValue(value.first), HashValue(value.second));
+  return static_cast<HashCode>(hasher);
 }
```

Comment:
These two overloads seem a bit limiting to me: we'll accept (for example) a pair where both elements are raw-hashable, or a pair where both elements are custom-hashable, but not a mixture.

I wonder if it'd be better for these to accept types that are hashable (in any way), excluding the cases handled by the above raw-only overloads? (Maybe `HasCustomCarbonHashValue<T>` should also be true in that case, given that it's invoking a custom function for part of the work.)